2017/gonumtour.slide

```diff
@@ -0,0 +1,306 @@
+
+- Hidden allocations
+- What are the types?
+- What does `'` do? Implicit? Copy? How to find out? Not easy to Google...
+
+Gonum sides with simplicity and consistency
+
+* "Productive, Performant, and Scalable"
+- Provide consistent APIs with design to interfaces
+- Allow memory reuse where reasonable (80/20 rule)
+- Many functions are concurrent-capable  ("opt-in" concurrency)
+
+	// Derivative estimates the derivative of the function f at the given location.
+	// The finite difference formula, the step size, and other options are
+	// specified by settings. If settings is nil, the first derivative will be
+	// estimated using the Forward formula and a default step size.
+	func Derivative(f func(float64) float64, x float64, settings *Settings) float64 {
+	...
+	}
+
+	// Settings is the settings structure for computing finite differences.
+	type Settings struct {
+		...
+		Concurrent bool // Should the function calls be executed concurrently.
+	}
+
+* Toward a Solution to the Two Language Problem
+- Pure Go - No mandatory C/Fortran
+- Translated 125+ LAPACK and BLAS functions
+- Competitive in speed (especially for full programs)
+
+* Gonum Design Philosophy
+
+* Panics vs. Errors
+- Rule #1: NO SILENT BUGS - Input is either clearly correct or it is wrong
+
+- Panic: Error is easy to check before call
+
+	// Add adds, element-wise, the elements of s and dst, and stores in dst.
+	// Panics if the lengths of dst and s do not match.
+	func Add(dst, s []float64) {
+		...
+	}
+
+- Error: Impossible to check without performing the operation
+
+	// Solve finds a minimum-norm solution to a system of linear equations defined
+	// by the matrices a and b. If A is singular or near-singular, a Condition error
+	// is returned.
+	func (m *Dense) Solve(a, b Matrix) error {
+		...
+	}
+
+* Memory Re-Use
+- If a function creates a type, allow a receiver as a first argument
+- But allow a nil receiver for ease
+
+	// Rand generates a random number according to the distributon.
+	// If the input slice is nil, new memory is allocated, otherwise the result is stored
+	// in place.
+	func (n *Normal) Rand(x []float64) []float64 {
+
+   
+	// CovarianceMatrix returns the covariance matrix calculated from a matrix of data, x.
+	//
+	// If cov is not nil it must either be zero-sized or have the same number of
+	// columns as the input data matrix. cov will be used as the destination for
+	// the covariance data. If cov is nil, a new mat.SymDense is allocated for
+	// the destination.
+	func CovarianceMatrix(cov *mat.SymDense, x mat.Matrix, weights []float64) *mat.SymDense {
+
+* "Functional" Programming
+- Functions and methods should have to modify state, or should not modify state
+- Even at the cost of extra allocations
+
+	// Reweight sets the weight of item idx to w. The input weight must be
+	// non-negative, and after reweighting at least one of the weights must be
+	// positive.
+	func (c Categorical) Reweight(idx int, w float64) {
```

Comment:
I just deleted the code snippet. It really needs two code snippets and a bunch more words to significantly improve the information content.